app/views.py

```diff
@@ -0,0 +1,164 @@
+    is_git_request = ('repo' in args and 'path' in args)
+    valid_request = xor(is_file_request, is_git_request)
+    if not valid_request:
+        return render_template('404.html')
+
+    hubauth = HubAuth()
+
+    # authenticate() returns either a username as a string or a redirect
+    redirection = username = hubauth.authenticate()
+    is_authenticated = isinstance(username, str)
+    if not is_authenticated:
+        values = []
+        for k, v in args.items():
+            if not isinstance(v, str):
+                v = '&path='.join(v)
+            values.append('%s=%s' % (k, v));
+        return render_template(
+            'landing.html',
+            authenticate_link=redirection.location,
+            download_links=util.generate_git_download_link(args),
+            query='&'.join(values))
+
+    # Start the user's server if necessary
+    if not hubauth.notebook_server_exists(username):
+        return redirect('/hub/home')
+
+    if not current_app.tracker[username]:
+        thread = Thread(
+            target=execute_request,
+            args=(current_app,
+                  current_app.app_context(),
+                  request.url,
+                  username,
+                  is_file_request,
+                  args),
+        )
+        current_app.tracker[username] = thread
+        util.emit_estimate_update(current_app.tracker)
+    else:
+        thread = None
+    return render_template(
+        'progress.html',
+        username=username,
+        reusing_thread=thread is None)
+
+
+def execute_request(app, context, url, username, is_file_request, args):
+    """Execute the request -- either a file load or a git pull.
+
+    This process is ideally run on a separate thread, as all methods in here
+    emit broadcasts to user-specific namespaces. In this case, the namespaces
+    are uniquely identified by the user's `username`.
+    """
+    with context:
+        try:
+            if is_file_request:
+                redirection = download_file_and_redirect(
+                    username=username,
+                    file_url=args['file'],
+                    config=current_app.config,
+                )
+            else:
+                redirection = pull_from_github(
+                    username=username,
+                    repo_name=args['repo'],
+                    paths=args['path'],
+                    config=current_app.config,
+                )
+
+            util.emit_redirect('/' + username, redirection)
+            app.tracker.pop(username)
+            util.emit_estimate_update(current_app.tracker)
+            return redirection
+        except BaseException as e:
+            app.tracker[username].url = url
+            app.tracker[username].error = e
+            util.emit_redirect(username, url_for('error', username=username))
+
+
+@current_app.route('/start/<string:username>')
+def start(username):
+    """Hack - GIL prevents Python from multi-threading.
```

Comment:
Think you can explain this a bit more? To my understanding, we initialize a thread in the landing view and use client-side JS to hit this endpoint. I'm not exactly sure how exactly the GIL limits multi-threading, but I'm assuming it prevents us from starting the thread in the landing method. Why is that the case?